### DISCUSSION TITLE
add new attribute into cue value

### DISCUSSION THREAD

**Discussion by ying-jeanne** [cue-lang/cue#1086] [Q&A]
Hi,

I am trying to search in the code base where i could actually read a Cue value, then if the value doesn't have an attribute, then add the attribute to the value.

But i didn't see any function for modifying those attributes, could you give an hint how i can do it?

**Comment by verdverm** [comment]
Can you clarify that you are trying to work with `@attributes` via the Go API?

From my understanding, Attributes are a read-only API currently.

**Reply by ying-jeanne** [reply]
according to what i see in the code, it seems like it is read-only, but i would need to add the attributes according to the conditions for the current project, i wonder is there any hack that i can do to make it happen?

**Reply by verdverm** [reply]
The best fix is probably to change the code around CUE so that this is not a hard need. For example, you could wrap the value with a Go struct which has the attribute information. Then you would fill that in from the CUE if possible and add any extras to the Go struct as needed.

It may be possible to maintain a fork of `cue` which allows modifying attributes, though that is a bit of a rabbit hole. There are always string manipulation hacks (i.e. writing templates, feeding the output to CUE, filling values)

**Comment by myitcv** [comment]
Per @eonpatapon on Slack, you can simply unify the value that is missing the attributes with one that has the attributes to create a new value that has those attributes:

```
go mod tidy
go run main.go

-- go.mod --
module mod.com

go 1.16

require cuelang.org/go v0.4.0
-- main.go --
package main

import (
	"fmt"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
)

func main() {
	ctx := cuecontext.New()

	x := ctx.CompileString("{5}")
	printVal("x", x)

	y := ctx.CompileString("{_,@x()}")
	printVal("y", y)

	z := x.Unify(y)
	printVal("z", z)
}

func printVal(name string, v cue.Value) {
	var attrNames []string
	for _, a := range v.Attributes(cue.ValueAttr) {
		attrNames = append(attrNames, a.Name())
	}
	fmt.Printf("%v: %#v, has attributes: %v\n", name, v, attrNames)
}
```

gives:

```
x: 5, has attributes: []
y: _, has attributes: [x]
z: 5, has attributes: [x]
```

**Reply by verdverm** [reply]
Does the act of creating a new value circumvent any of the normal attribute propagation rules?

For example, they don't normally propagate through conjunctions

```
$ cat extend.cue 
#extend: string @extended()
bar: string @myattr(bar)
bar: #extend

$ cue def -A extend.cue 
#extend: string           @extended()
bar:     #extend & string @myattr(bar)

$ cue eval -A extend.cue 
#extend: string @extended()
bar:     string @myattr(bar)
```

I wonder if you need to be careful with the Unify technique and placing the attribute "correctly" according to the propagation rules? And maybe this is perhaps why the API is read-only? (i.e. so that users don't shoot themselves in the foot by manually circumventing the rules that would break downstream tasks like Protobuf tags (iirc))

**Reply by ying-jeanne** [reply]
@myitcv I am trying to follow the unify solution, i might have forgot, but could you remind me if i have let's say a cue.Value, how i could print it out as string, for example
```
{
	fill?: "solid" | "dash" | "dot" | "square"
	dash?: [number]
}
```
to its exact cue presentation? so i can just add the @x in the end. i feel that i need to do it this way, since i am actually using a cue.iter to iterate a cue object and i would not know the iter.value() type in advance. and i need to add the attribute @cuetsy(targetType="interface") when it is absent. here is the code

```
iter, err := iCUE.Fields(cue.Definitions(true))
for iter.Next() {
    a := iter.Value().Attribute("cuetsy")
    _, found, err := a.Lookup(0, "targetType")
    if a.Err() != nil || !found || err != nil {
	    inst, err := rt.Compile("", `_,@cuetsy(targetType="interface")`)
	    if err != nil {
		    return iCUE, err
	    }
	    iCUE = iCUE.FillPath(cue.ParsePath(iter.Label()), inst.Value())
    }
    if err != nil {
	    return iCUE, err
    }
}
```
then i still didn't get changes on the attributes...
what is wrong here? whether i can solve it by getting the iter.Value cue string + @attribute then replace the origin value? then how i can get the print of iter.Value?

**Reply by myitcv** [reply]
@verdverm 

> Does the act of creating a new value circumvent any of the normal attribute propagation rules?

Not as far as I'm aware. Per https://github.com/cue-lang/cue/issues/894, the reason your example in https://github.com/cue-lang/cue/discussions/1086#discussioncomment-1000806 doesn't work as expected is that you are using field attributes, which do not propagate. Declaration attributes do:

https://gist.github.com/myitcv/2c4a95cbbe8d6f72eb8cde1b3781ed59

@ying-jeanne 

> how i could print it out as string to its exact cue presentation

I think you are after https://pkg.go.dev/cuelang.org/go@v0.4.0/cue#Value.Syntax and the https://pkg.go.dev/cuelang.org/go@v0.4.0/cue/format package.

> then i still didn't get changes on the attributes...

Can you share a full reproduction of what you are trying? We can take it from there.
